create_derivatives/routines.py

```diff
@@ -63,8 +66,72 @@ def get_as_uri(self, bag_filepath):
 
 
 class JP2Maker:
-    # TO DO: make JPG2000 derivatives
-    pass
+    """Creates JP2000 derivatives from TIFFs.
+
+    Creates JP2 directory in bag's data directory and JP2 derivatives.
+
+    Returns:
+        JP2000 derivatives in the bags' /data directory.
+        Updated bag process status.
+    """
+
+    def run(self):
+        bags_with_jp2s = []
+        for bag in Bag.objects.filter(process_status=Bag.PREPARED):
+            if Path.is_dir(bag.bag_path, "data", "service"):
+                tiff_files_dir = str(Path(bag.bag_path, "data", "service"))
+            else:
+                tiff_files_dir = str(Path(bag.bag_path, "data"))
+            self.jp2_path = self.create_jp2(bag, tiff_files_dir)
+            bag.process_status = Bag.JPG2000
+            bag.save()
+            bags_with_jp2s.append(bag.bag_identifier)
+        msg = "JPG2000s created." if len(bags_with_jp2s) else "No TIFF files ready for JP2 creation."
+        return msg, bags_with_jp2s
+
+    def calculate_layers(self):
+        """Calculates the number of layers based on pixel dimensions.
+        For TIFF files, image tag 256 is the width, and 257 is the height.
+        Args:
+            file (str): filename of a TIFF image file.
+        Returns:
+            layers (int): number of layers to convert to
+        """
+        # I can't use "file" here. What should I use?
+        with Image.open(file) as img:
+            width = [w for w in img.tag[256]][0]
+            height = [h for h in img.tag[257]][0]
+        return math.ceil((math.log(max(width, height)) / math.log(2)
+                          ) - ((math.log(96) / math.log(2)))) + 1
+
+    def create_jp2(self, bag, tiff_files_dir):
+        """Creates JPEG2000 files from TIFF files.
+        The default options for conversion below are:
+        - Compression ration of `1.5`
+        - Precinct size: `[256,256]` for first two layers and then `[128,128]` for all others
+        - Code block size of `[64,64]`
+        - Progression order of `RPCL`
+        """
+
+        default_options = ["-r", "1.5",
+                           "-c", "[256,256],[256,256],[128,128]",
+                           "-b", "64,64",
+                           "-p", "RPCL"]
+        tiff_files = matching_files(tiff_files_dir, prepend=True)
+        jp2_dir = Path(bag.bag_path, "data", "JP2")
+        if not jp2_dir.is_dir():
+            jp2_dir.mkdir()
+        # I feel like this is wrong. It doesn't make sense to use a for loop here, does it?
```

Comment:
Well, as written now this will create only one JPEG2000 file from all of the TIFF files, which is not what we want! I think your path definition in line 126 should be something more like `Path(jp2_dir, file)`, right?